source/ShipsFactory.cpp

```diff
@@ -0,0 +1,377 @@
+/* ShipsFactory.cpp
+Copyright (c) 2022 by Peter van der Meer
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "ShipsFactory.h"
+
+#include "DataNode.h"
+#include "Files.h"
+#include "GameData.h"
+#include "SpriteSet.h"
+#include "UniverseObjects.h"
+
+#include <algorithm>
+
+using namespace std;
+
+
+
+ShipsFactory::ShipsFactory(UniverseObjects &universe): universe(universe)
+{
+}
+
+
+
+// Load a ship from a datafile. Creation of the Ship object itself has
+// already been done by the caller. Getting only the DataNode as parameter
+// and returning a shared_ptr<Ship> is nicer than getting the ship as
+// parameter by reference, but the ES code allows overwriting ship-
+// definitions by a new load, so we need to support overwriting existing
+// ship definitions here.
```

Comment:
That is quite nice indeed.

Reply:
I thought about this a bit more (and tried to write some code for this); if we split Ships into a "ShipTemplate" (new class) and a "ShipRuntime" (the current Ship class), then we probably also need to split other classes like Mission into a "MissionTemplate" (a template that refers to ShipTemplates for to-be-instantiated NPCs) and a "MissionInstantiated" (an active mission that the player is actively doing, with NPCs being actual ships). Do you agree?

This PR here was originally created to separate the DataFile/SaveGame-loading from the runtime (and future template classes), so removing the GameData and UniverseObjects dependencies from the RunTime Ship.
I feel both the split of Ships into ShipRuntimes and  ShipTemplates and the removal of the GameData and UniverseObject dependencies from Runtimes and Templates (by making the load functions external) are valuable. What do you think?

Reply:
All `Load` members would eventually be split.

I want us to have a separate library that handles the loading of text files into templates so that we can share the game's parser and serializer implementations with the community.

So this library would convert DataNodes into some class-specific representation of the data, and handle writing that back into DataNodes.